### PR TITLE
export receiveany as method for raw req socket

### PR DESCRIPTION
**Pull request by tangsty**
export the implemention of receiveany as method for raw req socket

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -334,6 +334,41 @@ The raw TCP sockets returned by this module will contain the following extra met
 
 [Back to TOC](#directives)
 
+[tcpsock:receiveany](https://github.com/openresty/lua-nginx-module#tcpsockreceiveany)
+------------------
+
+**syntax:** *data, err = tcpsock:receiveany(max)*
+
+**context:** *rewrite_by_lua&#42;, access_by_lua&#42;, content_by_lua&#42;, ngx.timer.&#42;, ssl_certificate_by_lua&#42;, ssl_session_fetch_by_lua&#42;*
```
Comment (doujiang24) [inline README.md]: copy from http? there haven't such contexts.

File: t/162-raw-req-receiveany.t [hunk 2]
```diff
@@ -0,0 +1,118 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
```
Comment (doujiang24) [inline t/162-raw-req-receiveany.t]: better move this file to `067-req-socket.t`, and we need more tests, just like the `receiveany` tests in `058-tcp-socket.t`.

File: README.md [hunk 3]
```diff
@@ -334,6 +334,41 @@ The raw TCP sockets returned by this module will contain the following extra met
 
 [Back to TOC](#directives)
 
+[tcpsock:receiveany](https://github.com/openresty/lua-nginx-module#tcpsockreceiveany)
```
Comment (doujiang24) [inline README.md]: I think we should use `reqsock:receiveany` here.

File: README.md [hunk 4]
```diff
@@ -334,6 +334,41 @@ The raw TCP sockets returned by this module will contain the following extra met
...
+------------------
+
+**syntax:** *data, err = tcpsock:receiveany(max)*
+
+**context:** *rewrite_by_lua&#42;, access_by_lua&#42;, content_by_lua&#42;, ngx.timer.&#42;, ssl_certificate_by_lua&#42;, ssl_session_fetch_by_lua&#42;*
+
+Returns any data received by the connected socket, at most `max` bytes.
```
Comment (doujiang24) [inline README.md]: we may not need these (line 344 - 366) here, we can refer to http-lua doc instead.

File: README.md (hunk 3 above)
Comment (tangsty) [reply README.md]: The following are new methods for raw req socket. All use tcpsock

File: README.md (hunk 4 above)
Comment (tangsty) [reply README.md]: OK, removed, use a link to lua-nginx-module's tcpsock:receiveany

File: README.md (hunk 1 above)
Comment (tangsty) [reply README.md]: Cleaned up

File: README.md [hunk 5]
```diff
@@ -334,10 +334,23 @@ The raw TCP sockets returned by this module will contain the following extra met
...
 
-tcpsock:shutdown
+reqsock:receiveany
+------------------
+
+**syntax:** *data, err = reqsock:receiveany(max)*
+
+**context:** *content_by_lua&#42;, ngx.timer.&#42;, ssl_certificate_by_lua&#42;*
+
+This method's detail is same to [tcpsock:receiveany](https://github.com/openresty/lua-nginx-module#tcpsockreceiveany)
```
Comment (doujiang24) [inline README.md]: This method is similar to the [tcpsock:receiveany](https://github.com/openresty/lua-nginx-module#tcpsockreceiveany) method.

File: README.md [hunk 6]
```diff
@@ -175,7 +175,7 @@ purpose instead.
 
 **Note:** the lingering close directive that used to exist in older version of
 `stream_lua_nginx_module` has been removed and can now be simulated with the
-newly added [tcpsock:shutdown](#tcpsockshutdown) API if necessary.
+newly added [reqsock:shutdown](#reqsockshutdown) API if necessary.
```
Comment (doujiang24) [inline README.md]: changed by accident?

File: t/067-req-socket.t [hunk 7]
```diff
@@ -13,7 +13,7 @@ our $HtmlDir = html_dir;
 no_long_string();
 #no_diff();
 #log_level 'warn';
-no_shuffle();
+#no_shuffle();
```
Comment (doujiang24) [inline t/067-req-socket.t]: ditto.

**Comment by gebing** [comment]
need this function, pls merge it.

**Comment by doujiang24** [comment]
> need this function, pls merge it.

we will merge it after the new release it out in the near feature days, and it will be bundled in the next release that will be out in about June.

File: t/067-req-socket.t [hunk 8]
```diff
@@ -823,3 +823,174 @@ done
...
+        end
+
+        sock:send("send data after read side ")
+        sock:send(err)
+    }
+--- stream_response chomp
+send data after read side timeout
+--- error_log
+receiveany unexpected err: timeout
+
```
Comment (agentzh) [inline t/067-req-socket.t:996]: This blank line wouldn't exist if you use the [reindex script](https://github.com/openresty/openresty-devel-utils/blob/master/reindex) to process this `.t` file.

File: t/067-req-socket.t [hunk 9]
```diff
@@ -823,3 +823,174 @@ done
...
+                    ngx.log(ngx.ERR, 'failed to send after error ',err)
+                end
+
+                break
+            end
+            ngx.say(data)
+        end
+
+        sock:send("send data after read side ")
+        sock:send(err)
```
Comment (agentzh) [inline t/067-req-socket.t:990]: It's a good idea to always check the return values of the `send()` method calls for any unexpected errors (or expected ones).

File: t/067-req-socket.t [hunk 10]
```diff
@@ -823,3 +823,174 @@ done
...
+--- stream_server_config
+    content_by_lua_block {
+        local sock, err = ngx.req.socket(true)
+        if sock == nil then
+            ngx.log(ngx.ERR, 'failed to get raw req socket', err)
+            return
+        end
+        sock:settimeouts(500, 500, 500)
+
+        local data, err, bytes = nil, nil
```
Comment (agentzh) [inline t/067-req-socket.t:969]: I don't think we need to explicitly initialize newly declared local variables with `nil`.

**Comment by agentzh** [comment]
This branch should be rebased to the latest master. I'm seeing conflicts during the rebase:

```
agentzh@glass ~/git/stream-lua-nginx-module 0 (PR/184)$ g rb master
First, rewinding head to replay your work on top of it...
Applying: export ngx.req.socket:receiveany
Applying: revert the or 1.17.8.1rc0 中的更新
Applying: count + 1 for receiveany added
Applying: add receiveany doc copied from lua-nginx-module, mod version for stream-lua-nginx-module 0.0.8
Applying: add test for raw req socket receiveany
Applying: TLSv1 disabled, switch to TLSv1.1
Using index info to reconstruct a base tree...
M	t/129-ssl-socket.t
Falling back to patching base and 3-way merge...
Auto-merging t/129-ssl-socket.t
CONFLICT (content): Merge conflict in t/129-ssl-socket.t
error: Failed to merge in the changes.
Patch failed at 0006 TLSv1 disabled, switch to TLSv1.1
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

**Pull request merged**